### DISCUSSION TITLE
Cache control

### DISCUSSION THREAD

**Discussion by hitecSmartHome** [BCsabaEngine/svelteesp32#36] [Ideas]
@BCsabaEngine  Is caching possible somehow so the device can remember and load the cached page without the esp on the network? 

Currently the caching works but the browser asks the esp for the etag and compares the cached etag with the one from the esp. 

Until this communication is done, the browser does not load anything. 

I wonder if i can bypass this so the browser loads the cached page anyway and asks for the etag later.

This all must happen without https so i can not use service workers

**Comment by hitecSmartHome** [comment]
@BCsabaEngine

So the behaviour Iam looking for is that the browser should load the cached webpage and revalidate with the etag after. Or if the esp is not in the network it should load the cached pages.

Headers worth checking: must-revalidate, max-stale, max-age

**Comment by BCsabaEngine** [comment]
You ask a very interesting question! Yes, we are currently using the no-cache option because we were not prepared to serve special needs. (We thought of ESP as a constantly powered, fault-tolerant system).

As a bank web developer, I am aware of the principle of cache operation and its practical implementation. (Fortunately, there is no nginx proxy or CDN involved here, which would make the operation more complicated.) What I can suggest is that instead of the current no-cache, we work with an arbitrarily adjustable value, specified in seconds, and send it in the max-age header, e.g. 7200 (=2 hours). This can be overridden by the browser header (also called max-age), which indicates what value it accepts. We will send the smaller of the two. (The esp does not know for sure how much time it is, so we cannot generate a date (expires after 2024.12.30 00:00:00)).

It will be valid for all pages, modern JS frameworks (webpack, babel) create JS files with new names anyway, so excessive caching of them will not be a problem.

It's New Year's Eve, and after that I'm on leave due to medical intervention, but maybe I'll have the opportunity to help you with this.

In summary: an optionally adjustable max-age value could be a good service. I plan to implement this.

Happy New Year!

**Reply by hitecSmartHome** [reply]
Thank you very much for the fast reply!

I will write down my use-case for you so you can understand better.

I have built a webserver with PsychicHTTP on my esp32.
I want to reach my esp32 over the internet even if I'm not on the same network as my esp32.
For this I have built my own relay server with bun.
The esp32 connects to this relay server via wss.
Server authenticates this esp32 with a unique identifier. ( Will not let any esp32 to connect to it )
When the device ( in this case a mobile phone ) loads the webpage from the esp32 first time, and the user logs into, the esp32 sets
an Authorization header in localstorage for the user.
The user from now on can turn off it's wifi and switch to mobile network.
The web-app will recognize this change and from this point, makes any http request to the relay server.
The relay server then relays this packet to the appropirate esp32 via wss, the wss processes it as a normal http request and sends back a response via wss.
Server get's this message and answers to the http request with an http response.
So the webpage favors the local connection and switches to server connection if it must.

This works really great but I have one BIG flaw in this logic. 
The webpage can not be loaded from the same URL if the user is not on the same network as the esp32. 
So the user must load the webpage from the local network first ( and do not close the page after that )

The cache would prevent this because it would load the page from cache even if the esp32 is not there and the webpage will take controll of the request sending from there. ( it would automatically switch to server connection if it can't ping the esp32 locally )

The thing is that I still need the eTag caching since the webpage will still be on the esp32 and updates will come.

**Reply by hitecSmartHome** [reply]
![image](https://github.com/user-attachments/assets/bd0c48dc-c6fd-46ce-8737-89a4808e4d43)

**Reply by hitecSmartHome** [reply]
For this all to work as intented ( because it works if the page is already loaded ), I need a local cached offline page to be available on the user's device so it can be loaded even when the esp32 is not on the network

**Reply by BCsabaEngine** [reply]
Thank you for describing in detail what and how you do it. Unfortunately I can't get involved in your development, I can "only" provide support for the operation of svelteesp32. If the cache control modification solves the problem, then we will be good.

So I will make the modification, which can be useful not only for you, but also for others. Whether this will be suitable in the "bun + wss + multiple networks + diverse browsers" environment, unfortunately I don't know.

**Reply by BCsabaEngine** [reply]
It might be worth considering that the system running bun would _always_ serve the page, and if esp is available, it would update it from there. It would only proxy if there is something, otherwise bun would act as a cache.

**Reply by hitecSmartHome** [reply]
> It might be worth considering that the system running bun would _always_ serve the page, and if esp is available, it would update it from there. It would only proxy if there is something, otherwise bun would act as a cache.

Thank you for the suggestion but not possible. If the server gives the page it would be https and browsers dont allow mixed content. It can only work the other way around. Even if the esp32 hosts a https webserver, browsers considering self signed ca as mixed content also. Self signed is unsecure.

**Reply by hitecSmartHome** [reply]
I can do requests from http to https but not the other way around unfortunately. I understand that you dont want to be involved in this, I just wrote down so you can better understand the use case. Thank you very much

**Reply by hitecSmartHome** [reply]
In its simple form, we only need a cached page which the browser can load without esp32 on the network.

**Comment by BCsabaEngine** [comment]
Work in progress: https://github.com/BCsabaEngine/svelteesp32/pull/37

**Reply by hitecSmartHome** [reply]
I have tried to replace every occourence of no-cache with max-age=x ( 1 hour in sec ) but the browser refuses to cache the entire page for some reason. It can cache like 60% of it and it does not load further. Will wait for this commit and see what happens. Thank you very much for the feedback

**Reply by BCsabaEngine** [reply]
> I have tried to replace every occourence of no-cache with max-age=x ( 1 hour in sec ) but the browser refuses to cache the entire page for some reason. It can cache like 60% of it and it does not load further. Will wait for this commit and see what happens. Thank you very much for the feedback

I did a couple of tests (I'm trying it with all three web servers on a real esp32, I just had an S3 on the table). I saw that Chrome processes the max-age value nicely, and if I open pages while browsing the site and then go back to the main page, it fetches the values ​​from the cache in every case. I unplugged the ESP from my machine and the site worked.

This is not true when I tried to access an IP address that was not on the network in the address bar of the browser.

**Reply by BCsabaEngine** [reply]
> Work in progress: #37

v1.7.0 released: https://github.com/BCsabaEngine/svelteesp32/releases/tag/v1.7.0

**Reply by hitecSmartHome** [reply]
I see. Iam using mdns and not hostname. Will try it and report back!
